dynamiq/nodes/retrievers/retriever.py

```diff
@@ -92,31 +122,61 @@ def format_content(self, documents: list[Document], metadata_fields: list[str] |
             formatted_docs.append(formatted_doc)
         return "\n\n".join(formatted_docs)
 
-    def execute(self, input_data: RetrievalInputSchema, **_) -> dict[str, Any]:
+    def execute(
+        self, input_data: VectorStoreRetrieverInputSchema, config: RunnableConfig | None = None, **kwargs
+    ) -> dict[str, Any]:
         """Execute the retrieval tool.
 
         Args:
             input_data (dict[str, Any]): Input data for the tool.
+            config (RunnableConfig, optional): Configuration for the runnable, including callbacks.
             **kwargs: Additional keyword arguments.
 
         Returns:
             dict[str, Any]: Result of the retrieval.
         """
 
         logger.info(f"Tool {self.name} - {self.id}: started with INPUT DATA:\n{input_data.model_dump()}")
+        config = ensure_config(config)
+        self.reset_run_state()
+        self.run_on_node_execute_run(config.callbacks, **kwargs)
 
+        filters = input_data.filters or self.filters
+        top_k = input_data.top_k or self.top_k
+
+        alpha = input_data.alpha or self.alpha
+        query = input_data.query
         try:
-            text_embedder_output = self.text_embedder.run(input_data={"query": input_data.query})
+            kwargs = kwargs | {"parent_run_id": kwargs.get("run_id")}
+            text_embedder_output = self.text_embedder.run(
+                input_data={"query": query}, run_depends=self._run_depends, config=config, **kwargs
+            )
+            self._run_depends = [NodeDependency(node=self.text_embedder).to_dict()]
             embedding = text_embedder_output.output.get("embedding")
 
-            document_retriever_output = self.document_retriever.run(input_data={"embedding": embedding})
+            document_retriever_output = self.document_retriever.run(
+                input_data={
+                    "embedding": embedding,
+                    "top_k": top_k,
+                    "filters": filters,
+                    "alpha": alpha,
+                    **({"query": query} if alpha else {}),
+                },
+                run_depends=self._run_depends,
+                config=config,
+                **kwargs,
+            )
+            self._run_depends = [NodeDependency(node=self.document_retriever).to_dict()]
             retrieved_documents = document_retriever_output.output.get("documents", [])
             logger.debug(f"Tool {self.name} - {self.id}: retrieved {len(retrieved_documents)} documents")
 
             result = self.format_content(retrieved_documents)
             logger.info(f"Tool {self.name} - {self.id}: finished with RESULT:\n{str(result)[:200]}...")
 
-            return {"content": result}
+            if self.is_optimized_for_agents:
```

Comment:
Can we return both `result` and `documents`? I don't see limitations
`return {"content": 'result', "documents": retrieved_documents}`